README.md
# EH-WebComponents

These are some simple utility Web Components I find myself needing again and
again. They are all completely self-contained in individual `.min.js` files.
Though a bundle is also available.

These are intended to be dropped into sites without a build process. Reference
them directly in your HTML.

## Installation

Ideally, you don't install anything to use these. There is -- intentionally --
no npm package, no installer, no tooling required to use these components.

You can load individual components via CDN, i.e.:
```
<script src="https://cdn.jsdelivr.net/gh/tdesposito/eh-webcomponents/dist/eh-loading-modal.min.js"></script>
```

You can load ALL of them in a single bundle:
```
<script src="https://cdn.jsdelivr.net/gh/tdesposito/eh-webcomponents/dist/ehcomponents-bundle.min.js"></script>
```

You can copy them into your project, then serve them from there:
```
<script src="/static/eh-loading-modal.min.js"></script>
```

## Components Available
* [`<eh-confirm-modal>`](docs/confirm-modal.md): A full-screen modal confirm dialog box.
* [`<eh-loading-modal>`](docs/loading-modal.md): A full-screen "loading" modal pop-over.
* [`<eh-message-modal>`](docs/message-modal.md): A full-screen modal showing a message.
* [`<eh-notificaiton>`](docs/notification.md): Displays one or more notifications.
* [`<eh-toggle>`](docs/toggle.md): A single toggle switch.

## Styling Components
Each component uses CSS Custom Properties for styling. See each component's
documentation for a complete list.

These CSS Custom Properties are of the form:
```
--{aspect}[_{state}]-{css-property}: [{default value}]
```

In your stylesheets:
```
/* target a single instance of a component, by ID - as one does: */
#my_id_for_a_component {
  --box-color: 255 15 15;
  --button_hover-color: 0 255 185
}
/* or EVERY eh-message-modal on the page: */
eh-message-modal {
  --box-top: 100px;
}
/* You can target the slotted (child) content as usual. */
/* Here, you don't need the '--' since we're manipulating a child node: */
eh-message-modal pre {
  font-family: 'Comic Sans'
}
```
__*NOTE:* All colors are expressed as space-separated RGB values!__

## Demos

[See them in action](https://tdesposito.github.io/EH-WebComponents/)

## Contributing
I welcome comments, suggestions, and requests for components. Or build your own
and send me a pull request!

I intend that this library NOT become more complex to use or build:
* I don't want to adopt a framework.
* I don't want to create a framework.
* I don't want to create an NPM package.
* I don't want a complex build process. The [one in place](docs/dev.md) is super simple and, I think, Good Enough.

That said, if you want to create an NPM package, please let me know and we can talk.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [v1.2.0](https://github.com/tdesposito/EH-WebComponents/releases/tag/v1.2.0) - 2021-10-15

<small>[Compare with v1.1.0](https://github.com/tdesposito/EH-WebComponents/compare/v1.1.0...v1.2.0)</small>

### Added
- Add eh-inline-edit ([3d4e5c7](https://github.com/tdesposito/EH-WebComponents/commit/3d4e5c75443d8cc85e1d129f65f0e69fb5cfdc1a) by Todd Esposito).
- Add eh-notification ([4f95c13](https://github.com/tdesposito/EH-WebComponents/commit/4f95c1355c56bf347c2ca7a11e6834209a8ccb69) by Todd Esposito).
- Add eh-toggle component ([0733a03](https://github.com/tdesposito/EH-WebComponents/commit/0733a03a39220ad018282632f42fdbba56dec541) by Todd Esposito).

### Fixed
- Fix "onchange" handling ([07c61d1](https://github.com/tdesposito/EH-WebComponents/commit/07c61d1a3637b0dbdcdc46f347a2777cbbd1627b) by Todd Esposito).
- Fix css ordering for content-in-disabled ([52f9c23](https://github.com/tdesposito/EH-WebComponents/commit/52f9c2398366eded07788c30f97770a58f9e62eb) by Todd Esposito).

### Removed
- Remove obsolete "-{version}" from build/bundle process ([85c3031](https://github.com/tdesposito/EH-WebComponents/commit/85c303196e51b7e43e95e6820607e19ed785ca49) by Todd Esposito).


## [v1.1.0](https://github.com/tdesposito/EH-WebComponents/releases/tag/v1.1.0) - 2021-09-15

<small>[Compare with first commit](https://github.com/tdesposito/EH-WebComponents/compare/256552bd7cb8186c3f9e0e81f29b239182818daf...v1.1.0)</small>

### Added
- Add link to demos ([b4a4edb](https://github.com/tdesposito/EH-WebComponents/commit/b4a4edb34bb4fd4149ccee954da1b7c61cd89563) by Todd Esposito).
- Add redirect for github pages ([264461f](https://github.com/tdesposito/EH-WebComponents/commit/264461f0cbe539a2bcceebf72d6d11d5eb0e8879) by Todd Esposito).
- Add license ([357f71e](https://github.com/tdesposito/EH-WebComponents/commit/357f71e4be92e3b7db6730cd0073247a95f42379) by Todd Esposito).

### Fixed
- Fix loc for dev server ([9c235a9](https://github.com/tdesposito/EH-WebComponents/commit/9c235a96248c9a293735a4cd56bb25168bc5ed48) by Todd Esposito).

### Removed
- Remove incorrectly-added versioned dist files ([b78fa92](https://github.com/tdesposito/EH-WebComponents/commit/b78fa920b90699d5afcbc04e38c8dae797493988) by Todd Esposito).




src/eh-confirm-modal.js
// This file is part of EH-WebComponents, Copyright (C) Todd D. Esposito 2021.
// Distributed under the MIT License (see https://opensource.org/licenses/MIT).

window.customElements.define('eh-confirm-modal',
  class EhConfirmModal extends HTMLElement {
    constructor() {
      super()
      this.attachShadow({ mode: "open" })

      const visible = this.getAttribute('visible') || ""

      this.shadowRoot.innerHTML = `<!-- ##HTML -->
        <style>
          .wrapper {
            background-color: rgb(var(--mask-background-color, 198 187 187));
            height: 100%;
            left: 0;
            opacity: 0;
            position: fixed;
            top: 0;
            visibility: hidden;
            width: 100%;
            z-index: 1;
          }
          .visible {
            opacity: var(--mask-opacity, 97%);
            visibility: visible;
            transition: visibility 0s linear 0s, opacity var(--transition-duration, .25s) 0s;
          }
          .modal {
            background-color: rgb(var(--box-background-color, 254 254 254));
            border-radius: var(--box-border-radius, 10px);
            border: var(--box-border, 1px solid rgba(var(--box-background-color, 254 254 254) / 10%));
            box-shadow: 0 4px 8px 0 rgba(0 0 0 / 20%), 0 6px 20px 0 rgba(0 0 0 / 19%);
            left: var(--box-left, 15%);
            min-width: var(--box-width, 70%);
            padding: var(--box-padding, 12px);
            position: relative;
            text-align: var(--box-text-align, center);
            top: var(--box-top, 20%);
            width: var(--box-width, 70%);
            z-index: 10;
          }
          .title {
            background-color: rgb(var(--title-background-color, 86 120 229));
            border-radius: var(--title-border-radius, 6px);
            border: var(--title-border, 1px solid rgba(var(--title-background-color, 86 120 229) / 50%));
            color: rgb(var(--title-color, 255 255 255));
            font-size: var(--title-font-size, 1.5rem);
            padding: var(--title-padding, 2px 16px);
          }
          .content {
            margin: var(--content-margin, 20px 2px);
            padding: var(--content-padding, 2px 16px);
          }
          .button-container {
            text-align: center;
          }
          button {
            background-color: rgba(var(--button-background-color, 27 149 9) / 95%);
            border-radius: var(--button-border-radius, 6px);
            border: var(--button-border, 1px solid rgba(var(--button-background-color, 27 149 9) / 50%));
            color: rgb(var(--button-color, 255 255 255));
            cursor: pointer;
            font-size: var(--button-font-size, 1.1rem);
            line-height: var(--button-line-height, 16px);
            margin: var(--button-margin, 6px);
            min-width: var(--button-min-width, 10rem);
            padding: var(--button-padding, 17px 34px);
            transition: opacity var(--button-transition-duration, var(--transition-duration, .25s)) 0s;
          }
          button.confirm {
            background-color: rgba(var(--confirm-background-color, var(--button-background-color, 27 149 9)) / 95%);
          }
          button.cancel {
            background-color: rgba(var(--cancel-background-color, var(--button-background-color, 27 149 9)) / 95%);
          }
          button:hover {
            transform: var(--button_hover-transform, scale3d(1.02, 1.02, 1));
          }
          button.confirm:hover {
            background-color: rgba(var(--confirm-background-color, var(--button-background-color, 27 149 9)) / 100%);
          }
          button.cancel:hover {
            background-color: rgba(var(--cancel-background-color, var(--button-background-color, 27 149 9)) / 1005%);
          }
        </style>
        <div class="wrapper ${visible}">
          <div class='modal'>
            <div class='title'><slot name="title">Confirm</slot></div>
            <div class='content'>
              <slot>Are You Sure You Want To Continue?</slot>
            </div>
            <div class='button-container'>
              <button class='confirm'><slot name="confirm-text">Confirm</slot></button>
              <button class='cancel'><slot name="cancel-text">Cancel</slot></button>
            </div>
          </div>
        </div>
      <!-- ##/HTML -->`

      this.wrapper = this.shadowRoot.querySelector(".wrapper")
      this.confirmButton = this.shadowRoot.querySelector(".confirm")
      this.cancelButton = this.shadowRoot.querySelector(".cancel")

      this.onConfirm = this.onConfirm.bind(this)
      this.onCancel = this.onCancel.bind(this)
    }

    onConfirm() {
      this.removeAttribute("visible")
      if (this.getAttribute("onConfirm")) {
        let fn
        try {
          fn = Function(this.getAttribute("onConfirm"))
          fn()
        } catch (error) {
          console.error("eh-confirm-modal: can't execute the 'onConfirm' routine.")
        }
      }
      this.dispatchEvent(new CustomEvent("confirm"))
    }

    onCancel() {
      this.removeAttribute("visible")
      this.dispatchEvent(new CustomEvent("cancel"))
    }

    connectedCallback() {
      this.confirmButton.addEventListener("click", this.onConfirm)
      this.cancelButton.addEventListener("click", this.onCancel)
    }

    disconnectedCallback() {
      this.confirmButton.removeEventListener("click", this.onConfirm)
      this.cancelButton.removeEventListener("click", this.onCancel)
    }

    get visible() {
      return this.hasAttribute("visible")
    }
    set visible(v) {
      if (v) {
        this.setAttribute("visible", "")
      } else {
        this.removeAttribute("visible")
      }
    }

    static get observedAttributes() {
      return ["visible"]
    }

    attributeChangedCallback(name, oldValue, newValue) {
      switch (name) {
      case 'visible':
        if (newValue === null) {
          this.wrapper.classList.remove("visible")
        } else {
          this.wrapper.classList.add("visible")
        }
      }
    }

    show() { this.setAttribute("visible", "") }
    hide() { this.removeAttribute("visible") }
  }
)
